File: apps/ios/Maverick/Model/MavCoreAnalyticsRuntime.swift
import Foundation

/// The generated core, in the shape `MavAnalyticsEngine` asks for.
///
/// Everything here is translation and nothing is a decision: the plan, the ordering, the reasons
/// and the cache all come from Rust. The one thing this file does is turn the core's strings into
/// the Swift types the reducer switches on, and it throws on an unrecognised one rather than
/// defaulting — a new unavailable reason should surface as a loud failure, not as a card that
/// silently renders as "working".
struct MavCoreAnalyticsRuntime: MavAnalyticsRuntime {
  let runtime: MavRuntime

  func host() -> MavModelBridge.Host { runtime }

  func admitPPGStages(deviceID: UInt64, atMs: Int64) throws {
    _ = try runtime.admitPpgStages(deviceId: deviceID, atMs: atMs)
  }

  func plan(
    deviceID: UInt64,
    atMs: Int64,
    mode: MavAnalyticsEngine.RunMode,
    profileFields: [String]
  ) throws -> [MavPlannedStage] {
    try runtime.analyticsPlan(
      deviceId: deviceID,
      atMs: atMs,
      mode: mode.rawValue,
      profileFields: profileFields
    ).stages.map(Self.stage(from:))
  }

  func profileFields() -> [String] { (try? runtime.wearerProfileFields()) ?? [] }

  func cacheCompletedAt() throws -> [String: Int64] {
    var out: [String: Int64] = [:]
    for entry in try runtime.analyticsCache() { out[entry.modelSlug] = entry.completedAtMs }
    return out
  }

  static func stage(from report: ModelStageReport) throws -> MavPlannedStage {
    guard let state = MavStageState(rawValue: report.state) else {
      throw MavModelError.unknown("the core reported an unknown stage state \(report.state)")
    }
    return MavPlannedStage(
      model: report.modelSlug,
      signal: report.signal,
      state: state,
      displayable: report.displayable,
      unavailable: try unavailable(from: report)
    )
  }

  private static func unavailable(from report: ModelStageReport) throws -> MavUnavailable? {
    switch report.unavailableReason {
    case .none: return nil
    case "missing_streams": return .missingStreams(report.missingStreams)
    case "missing_profile": return .missingProfile(report.missingProfile)
    case "upstream_unavailable": return .upstreamUnavailable(report.blockingModel ?? "")
    case "preprocessing_not_ported":
      return .preprocessingNotPorted(report.missingPreprocessing ?? "")
    case .some(let other):
      throw MavModelError.unknown("the core reported an unknown unavailable reason \(other)")
    }
  }
}

/// A core response this build does not understand.
enum MavModelError: Error {
  case unknown(String)
}

File: apps/android/app/src/main/java/com/sennnen/mav/ml/MavCoreAnalyticsRuntime.kt
package com.sennnen.mav.ml

import uniffi.mav_ffi.MavRuntime
import uniffi.mav_ffi.ModelInferenceRequest
import uniffi.mav_ffi.ModelInferenceResult
import uniffi.mav_ffi.ModelStageReport
import uniffi.mav_ffi.ModelTensor

/**
 * The generated core, in the shape [MavAnalyticsEngine] asks for.
 *
 * Everything here is translation and nothing is a decision: the plan, the ordering, the reasons
 * and the cache all come from Rust. The one thing this file is allowed to do is turn the core's
 * strings into the Kotlin types the reducer switches on, and it fails loudly on an unrecognised
 * one rather than defaulting — a new unavailable reason should show up as a compile-adjacent
 * bug, not as a card that silently renders as "working".
 */
class MavCoreAnalyticsRuntime(private val runtime: MavRuntime) : MavAnalyticsRuntime {

    override fun host(): MavModelBridge.Host = object : MavModelBridge.Host {
        override fun nextModelInference(): ModelInferenceRequest? = runtime.nextModelInference()

        override fun submitModelInference(
            requestId: ULong,
            outputs: List<ModelTensor>,
            modelSha256: String,
            completedAtMs: Long,
        ): ModelInferenceResult =
            runtime.submitModelInference(requestId, outputs, modelSha256, completedAtMs)

        override fun cancelModelInference(requestId: ULong): Boolean =
            runtime.cancelModelInference(requestId)
    }

    override fun admitPpgStages(deviceId: ULong, atMs: Long) {
        runtime.admitPpgStages(deviceId, atMs)
    }

    override fun plan(
        deviceId: ULong,
        atMs: Long,
        mode: MavRunMode,
        profileFields: List<String>,
    ): MavPlan {
        val report = runtime.analyticsPlan(deviceId, atMs, mode.wire, profileFields)
        return MavPlan(
            stages = report.stages.map(::stageOf),
            coverage = report.coverage.associate {
                it.signal to MavSignalCoverage(it.total.toInt(), it.runnable.toInt())
            },
        )
    }

    override fun profileFields(): List<String> = runtime.wearerProfileFields()

    override fun cacheCompletedAt(): Map<String, Long> =
        runtime.analyticsCache().associate { it.modelSlug to it.completedAtMs }

    private fun stageOf(report: ModelStageReport): MavPlannedStage = MavPlannedStage(
        model = report.modelSlug,
        signal = report.signal,
        state = when (report.state) {
            "ready" -> MavStageState.READY
            "blocked" -> MavStageState.BLOCKED
            "cached" -> MavStageState.CACHED
            "unavailable" -> MavStageState.UNAVAILABLE
            else -> throw MavModelException("the core reported an unknown stage state ${report.state}")
        },
        displayable = report.displayable,
        unavailable = unavailableOf(report),
    )

    private fun unavailableOf(report: ModelStageReport): MavUnavailable? =
        when (report.unavailableReason) {
            null -> null
            "missing_streams" -> MavUnavailable.MissingStreams(report.missingStreams)
            "missing_profile" -> MavUnavailable.MissingProfile(report.missingProfile)
            "upstream_unavailable" ->
                MavUnavailable.UpstreamUnavailable(report.blockingModel.orEmpty())
            "preprocessing_not_ported" ->
                MavUnavailable.PreprocessingNotPorted(report.missingPreprocessing.orEmpty())
            else -> throw MavModelException(
                "the core reported an unknown unavailable reason ${report.unavailableReason}",
            )
        }
}
